pkg/controllers/deprovisioning/controller.go

```diff
@@ -411,57 +421,56 @@ func (c *Controller) waitForDeletion(ctx context.Context, node *v1.Node) {
+	// cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := c.setNodesUnschedulable(ctx, true, nodeNamesToRemove...); err != nil {
+		return fmt.Errorf("cordoning nodes, %w", err)
 	}
-	oldNode := action.nodesToRemove[0]
 
-	// cordon the node before we launch the replacement to prevent new pods from scheduling to the node
-	if err := c.setNodeUnschedulable(ctx, action.nodesToRemove[0].Name, true); err != nil {
-		return fmt.Errorf("cordoning node %s, %w", oldNode.Name, err)
-	}
-
-	nodeNames, err := c.provisioner.LaunchNodes(ctx, provisioning.LaunchOptions{RecordPodNomination: false}, action.replacementNode)
+	nodeNames, err := c.provisioner.LaunchNodes(ctx, provisioning.LaunchOptions{RecordPodNomination: false}, action.replacementNodes...)
 	if err != nil {
-		// uncordon the node as the launch may fail (e.g. ICE or incompatible AMI)
-		err = multierr.Append(err, c.setNodeUnschedulable(ctx, oldNode.Name, false))
+		// uncordon the nodes as the launch may fail (e.g. ICE or incompatible AMI)
+		err = multierr.Append(err, c.setNodesUnschedulable(ctx, false, nodeNamesToRemove...))
 		return err
 	}
-	if len(nodeNames) != 1 {
-		// shouldn't ever occur as we are only launching a single node
-		return fmt.Errorf("expected a single node name, got %d", len(nodeNames))
+	if len(nodeNames) != len(action.replacementNodes) {
+		// shouldn't ever occur since a partially failed LaunchNodes should return an error
+		return fmt.Errorf("expected %d node names, got %d", len(action.replacementNodes), len(nodeNames))
 	}
-
-	metrics.NodesCreatedCounter.WithLabelValues(metrics.DeprovisioningReason).Inc()
-
-	// We have the new node created at the API server so mark the old node for deletion
-	c.cluster.MarkForDeletion(oldNode.Name)
-
-	var k8Node v1.Node
-	// Wait for the node to be ready
-	var once sync.Once
-	if err := retry.Do(func() error {
-		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodeNames[0]}, &k8Node); err != nil {
-			return fmt.Errorf("getting node, %w", err)
-		}
-		once.Do(func() {
-			c.recorder.Publish(deprovisioningevents.LaunchingNode(&k8Node, action.String()))
-		})
-
-		if _, ok := k8Node.Labels[v1alpha5.LabelNodeInitialized]; !ok {
-			// make the user aware of why deprovisioning is paused
-			c.recorder.Publish(deprovisioningevents.WaitingOnReadiness(&k8Node))
-			return fmt.Errorf("node is not initialized")
+	metrics.NodesCreatedCounter.WithLabelValues(metrics.DeprovisioningReason).Add(float64(len(nodeNames)))
+
+	// We have the new nodes created at the API server so mark the old nodes for deletion
+	c.cluster.MarkForDeletion(nodeNamesToRemove...)
+	// Wait for nodes to be ready
+	// TODO @njtran: Allow to bypass this check for certain deprovisioners
+	for _, nodeName := range nodeNames {
+		var k8Node v1.Node
+		// Wait for the node to be ready
+		var once sync.Once
+		if err := retry.Do(func() error {
```

Comment:
We should probably publish all node events at once rather than trickle the events for each node one at-a-time

Reply:
Good idea, parallelizing the check now.